### PR TITLE
ENH: assert_array_less should report max violations instead of max differences 

### LINKED ISSUE #23894: ENH: `assert_array_less` should report max violations instead of max differences
### Proposed new feature or change:

When `numpy.testing.assert_array_less` fails, it reports a "Max absolute difference" and a "Max relative difference". These metrics are the max element-wise differences between the two tested arrays. These make sense to report when testing array equality, but they don't seem useful when asserting that one array is element-wise less than another. It would make more sense to report the maximum values by which the less-than constraint is violated. "Max absolute violation" could be calculated as:

```
np.maximum(array1 - array2, 0).max()
```

when `npt.assert_array_less(array1, array2)` fails. "Max relative violation" could be calculated similarly.

I'd be happy to implement this if people agree. Thanks!

### PR DESCRIPTION
**Pull request by undermyumbrella1**
Fixes #23894

### CODE REVIEW

**Comment by ngoldbaum** [comment]
Looks like some docstrings need to be updated, the test failures are real.

**Comment by ellaella12** [comment]
ok updated, is the travis ci pull request build failure an issue on my end too?

**Comment by ngoldbaum** [comment]
Nope, that one's unrelated. The Travis PPC64LE runners have been flaky for quite a while
now.

File: numpy/testing/tests/test_utils.py
Comment (ngoldbaum) [inline numpy/testing/tests/test_utils.py:97]: nit, here and below: you can pass a `match=some_regex` to `pytest.raises` to check for a matching message.

File: numpy/testing/tests/test_utils.py
Comment (ngoldbaum) [inline numpy/testing/tests/test_utils.py:219]: why is this commented out?

**Review by ngoldbaum — COMMENTED** [review]
I think this is a clear improvement, thank you for working on this, it will make the debugging experience for the entire PyData ecosystem nicer when a test using this assert fails.

The implementation looks correct and the new tests make sense. I think you can substantially slim down the new tests you've added by checking for a matching message directly in `pytest.raises`, see comments below.

File: numpy/testing/tests/test_utils.py
Comment (ellaella12) [reply numpy/testing/tests/test_utils.py]: Thank you for the review, i have updated the error message checks to use match

**Comment by ellaella12** [comment]
Hi, any updates on this pr? thank you

**Review by ngoldbaum — APPROVED** [review]

**Comment by ngoldbaum** [comment]
I added the triage review label so hopefully this will get a look at the next triage meeting unless someone else merges in the meantime. This is a lot of code so I want another set of eyes before merging.

**Comment by ellaella12** [comment]
ok, thank you

**Comment by ngoldbaum** [comment]
It looks like there's a conflict that needs to be fixed

**Comment by ngoldbaum** [comment]
Thanks @ellaella12!

**Pull request merged**
